### PR TITLE
LLM Validator

### PR DESCRIPTION
**Pull request by GWSzeto**
This is an implementation of the LLM validator. It ports over the existing implementation from the python side of instructor.

Few caveats to think about:
- llmValidator can only be used through the zod `transform` method. This is to be able to support both validation and modification of the value, similar to pydantics annotate functionality
- since the llm validator uses async/await, `asyncParse` needs to be used in this context
- Wrestled a bit with the prompt for the llm validator so that it was able to conform to the validation schema passed in for the validator schema

### CODE REVIEW

**Comment by GWSzeto** [comment]
@jxnl did you have issues with the llm validator not conforming to the schema before? Or am I just messing something up here?

File: src/dsl/validators.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      messages: [
+        {
+          role: "system",
+          content: `# MISSION
+You are a world class validation model. Capable to determine if the value provided is valid based on the statement given. If it is not, give a reason why and suggest a new value.
+        `
+        },
+        {
+          role: "user",
+          content: `Does value ${v} follow the statement: ${statement}`
```
Comment (jxnl) [inline src/dsl/validators.ts:48]: add some escape quotes to avoid some weird confusion

File: src/dsl/validators.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+  //eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async function llm(v: any, ctx: z.RefinementCtx) {
+    const response = await client.chat.completions.create({
+      response_model: Validator,
+      messages: [
+        {
+          role: "system",
+          content: `# MISSION
+You are a world class validation model. Capable to determine if the value provided is valid based on the statement given. If it is not, give a reason why and suggest a new value.
```
Comment (jxnl) [inline src/dsl/validators.ts:43]: mission?

File: tests/validators.test.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+          openaiClient: client
+        })
+      )
+    })
+    const questionAnswerValue = {
+      question: "What is the meaning of life?",
+      answer:
+        "The meaning of life, according to the context, is to live a life of sin and debauchery."
+    }
+    const { success, data } = (await ModifiedQuestionAnswer.safeParseAsync(
```
Comment (jxnl) [inline tests/validators.test.ts:65]: should instructor always use safeParseAsync?

File: src/dsl/validators.ts (hunk 2 above)
Comment (GWSzeto) [reply src/dsl/validators.ts:43]: Updated the prompt to try to get some of the tests working properly. It's based off of these templates that I found 
https://github.com/daveshap/ChatGPT_Custom_Instructions/blob/main/academic_copy_reviser.md

File: tests/validators.test.ts (hunk 3 above)
Comment (GWSzeto) [reply tests/validators.test.ts:65]: It needs to use safeParseAsync/parseAsync when an async/await function is used in either a refine or transform validation method.

Not sure if there's a consequence to using it on sync cases

Comment (jxnl) [reply tests/validators.test.ts:65]: yeah just thinking about what happens if we do

create(..., response_model:Model)

**Pull request closed without merging**